src/expr/src/relation/canonicalize.rs

```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
+    // Seed `predicates_to_apply` with the `Vec` of all predicates.
+    std::mem::swap(&mut predicates_to_apply, predicates);
+
+    while let Some(predicate_to_apply) = predicates_to_apply.pop() {
+        let mut replace_subexpr_other_predicates =
+            |expr: &MirScalarExpr, constant_bool: &MirScalarExpr| {
+                // Do not replace subexpressions equal to `expr` if `expr` is a
+                // literal to avoid infinite looping.
+                if !expr.is_literal() && !expr.typ(input_type).nullable {
+                    let replace_subexpr_other_predicate = |other_predicate: &mut MirScalarExpr| {
+                        let mut changed = false;
+                        visit_mut_replace_subexpr(
+                            other_predicate,
+                            &mut |e| replace(e, expr, constant_bool),
+                            &mut changed,
+                        );
+                        if changed {
+                            other_predicate.reduce(input_type);
+                        }
+                        changed
+                    };
+                    for other_predicate in predicates_to_apply.iter_mut() {
+                        replace_subexpr_other_predicate(other_predicate);
+                    }
+                    for other_idx in (0..applied_predicates.len()).rev() {
+                        if replace_subexpr_other_predicate(&mut applied_predicates[other_idx]) {
+                            predicates_to_apply.push(applied_predicates.remove(other_idx));
+                        }
+                    }
+                }
+            };
+
+        if let MirScalarExpr::CallUnary {
+            func: UnaryFunc::Not,
+            expr,
+        } = &predicate_to_apply
+        {
+            replace_subexpr_other_predicates(
+                expr,
+                &MirScalarExpr::literal_ok(Datum::False, ScalarType::Bool),
+            )
+        } else {
+            replace_subexpr_other_predicates(
+                &predicate_to_apply,
+                &MirScalarExpr::literal_ok(Datum::True, ScalarType::Bool),
+            );
+        }
+        applied_predicates.push(predicate_to_apply);
+    }
+    // Remove any predicates that have been reduced to "true"
+    applied_predicates.retain(|p| !p.is_literal_true());
+    *predicates = applied_predicates;
+
+    // 4) Sort and dedup predicates.
     predicates.sort();
     predicates.dedup();
 }
+
+/* #region helper functions for substituting a bool literal for a matching subexpression */
+
+fn replace(
+    predicate: &mut MirScalarExpr,
+    to_replace: &MirScalarExpr,
+    replace_with: &MirScalarExpr,
+) -> bool {
+    if predicate == to_replace {
+        *predicate = replace_with.clone();
+        return true;
+    }
+    false
+}
+
+fn visit_mut_replace_subexpr<F>(e: &mut MirScalarExpr, f: &mut F, changed: &mut bool)
+where
+    F: FnMut(&mut MirScalarExpr) -> bool,
```

Comment:
Some doc comments on this would really help! Or perhaps `replace_subexprs_not_cond` as a name? Would this be valuable in `MirScalarExpr`? I know there is very similar logic in `column_knowledge.rs`, and it might make sense to promote to "a thing one shouldn't try and do by hand". No need to do it in this PR, but sounding out the idea to see if it makes sense.

Reply:
I want to create a common method in `MirScalarExpr` for this because I know of similar code in the `JoinInputMapper`, but 1) I was afraid of the common method creating confusion because there would be changes to completely unrelated parts of the code. 2) I suspected that there were other parts of the code with similar logic that I did not know about, and I didn't want to spec out a common method until I saw those other parts. If you know of other places besides `column_knowledge.rs` and `join_input_mapper.rs` with similar logic, please let me know!